core/src/main/scala/org/apache/spark/storage/BlockObjectWriter.scala

```diff
@@ -203,12 +208,32 @@ private[spark] class DiskBlockObjectWriter(
     }
   }
 
-  override def write(value: Any) {
+  override def write(key: Any, value: Any) {
+    if (!initialized) {
+      open()
+    }
+
+    objOut.writeKey(key)
+    objOut.writeValue(value)
+    numRecordsWritten += 1
```

Comment:
To avoid code duplication, I think we could have called `recordWritten()` here instead, which would handle updating the bytes written etc.